### DISCUSSION TITLE
Is there any way to implement the following configuration in alfred?

### DISCUSSION THREAD

**Discussion by alkene5** [johnlindquist/kit#1279] [Q&A] [answered]
<img width="558" alt="image" src="https://github.com/johnlindquist/kit/assets/61105068/86fcf110-bd43-43f0-80ef-8b416c4bc5f6">
I have a lot of scripts that are making requests to some api, and I don't want those requests to be so fast that the server limits the frequency of access.

**Comment by johnlindquist** [comment]
Can you show me some of the code that's invoking the API? Is it invoking the API while you're typing?

**Reply by alkene5** [reply]
// inputChangeFunc is called whenever user input changes
// However, if the user types very fast, such as 'test, four requests will be sent in a very short time
// alfred provides a setting where inputChangeFunc is executed only if the user does not enter it again within 200 milliseconds
// In fact, it seems to me that `setTimeout` is a similar implementation, but I wonder if you would consider adding it to the standard library
```js
function inputChangeFunc(input) {
    return get('http://xxxx..')
}
let value = await arg('Input:', inputChangeFunc)
```

**Reply by alkene5** [reply]
```js
// Create a new function that doesn't fire frequently
function debounce(func, delay) {
    let timer = null;
    return function() {
        clearTimeout(timer);
        timer = setTimeout(() => {
            func.apply(this, arguments);
        }, delay);
    };
}

function doSomething() {
    console.log('Doing something...');
}

const debouncedDoSomething = debounce(doSomething, 1000);

debouncedDoSomething();     // No Exec
debouncedDoSomething();     // No Exec
setTimeout(debouncedDoSomething, 500);  // No Exec
setTimeout(debouncedDoSomething, 500 + 900);  // Exec, Because the next request is not called until more than 1 second later
setTimeout(debouncedDoSomething, 500 + 900 + 1100);  // Exec, The last call is going to be executed anyway
```
I probably used chatgpt to generate the code I needed, which could be done, but it was cumbersome

**Reply by johnlindquist** [reply, marked as answer]
Ah, ok, you need an `async` function. Script Kit will detect that it's "async" and automatically debounce it:

```js
await arg("Testing Async Function", async input => {
  // This will be debounced
  return input
})

// control the debounce
await arg(
  {
    placeholder: "Testing Async Function",
    debounceInput: 500,
  },
  async input => {
    // This will be debounced
    return input
  }
)

```

**Reply by alkene5** [reply]
Thank you. I hope the official documents can follow up these contents as soon as possible. I like this project very much😃
